File: 2021/07/18/str_str.py
"""
https://leetcode.com/problems/implement-strstr/
Author: Andrew Jarombek
Date: 7/18/2021
"""


def str_str(haystack: str, needle: str) -> int:
    if len(needle) == 0:
        return 0

    for i in range(len(haystack)):
        if needle == haystack[i:i + len(needle)]:
            return i

    return -1


if __name__ == '__main__':
    assert str_str("", "") == 0
    assert str_str("abc", "d") == -1
    assert str_str("hello", "ll") == 2

File: 2022/02/13/cpp/strStr.cpp
// https://leetcode.com/problems/implement-strstr/
// Author: Andrew Jarombek
// Date: 2/13/2022

#include <string>
#include <iostream>
#include <cassert>

#include "searchInsertPosition.h"
#include "maxSubArray.h"
#include "lengthOfLastWord.h"

using namespace std;

int strStr(string haystack, string needle) {
    if (needle.length() == 0) {
        return 0;
    }

    if (haystack.length() < needle.length()) {
        return -1;
    }

    for (int i = 0; i <= haystack.length() - needle.length(); i++) {
        cout << haystack.substr(i, needle.length()) << "\n";

        if (needle == haystack.substr(i, needle.length())) {
            return i;
        }
    }

    return -1;
}

int main() {
    assert(strStr("hello", "ll") == 2);
    assert(strStr("", "") == 0);

    vector<int> v {1, 3, 5, 7};
    assert(searchInsert(v, 2) == 1);

    vector<int> vec {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    assert(maxSubArray(vec) == 6);

    assert(lengthOfLastWord("Hello World") == 5);
}
